pkg/scheduler/controller/backupbucket/scheduler_reconciler.go

```diff
@@ -0,0 +1,229 @@
+// 6. Return error if none of the above step found seed.
+func (r *reconciler) determineSeed(backupBucket *gardencorev1alpha1.BackupBucket) (*gardenv1beta1.Seed, error) {
+	var (
+		candidatesWithMathingProfile    = make([]*gardenv1beta1.Seed, 0)
+		candidatesWithoutMathingProfile = make([]*gardenv1beta1.Seed, 0)
+		cloudProfileCache               = make(map[string]gardenv1beta1.CloudProvider) //cloudProfile name verses cloudProvider map
+	)
+
+	seeds := &gardenv1beta1.SeedList{}
+	if err := r.client.List(r.ctx, seeds); err != nil {
+		return nil, err
+	}
+
+	if len(seeds.Items) == 0 {
+		return nil, fmt.Errorf("no seed found for scheduling")
+	}
+	for _, seed := range seeds.Items {
+		if seed.DeletionTimestamp != nil || verifySeedAvailability(&seed) {
+			continue
+		}
+
+		provider, ok := cloudProfileCache[seed.Spec.Cloud.Profile]
+		if !ok {
+			cloudProfile := &gardenv1beta1.CloudProfile{}
+			if err := r.client.Get(r.ctx, kutil.Key(seed.Spec.Cloud.Profile), cloudProfile); err != nil {
+				return nil, err
+			}
+
+			provider, err := gardenhelper.DetermineCloudProviderInProfile(cloudProfile.Spec)
+			if err != nil {
+				return nil, err
+			}
+			cloudProfileCache[cloudProfile.Name] = provider
+		}
+		if backupBucket.Spec.Provider == string(provider) {
+			if seed.Spec.Cloud.Region == backupBucket.Spec.Region {
+				return &seed, nil
+			}
+			candidatesWithMathingProfile = append(candidatesWithMathingProfile, &seed)
+		}
+		candidatesWithoutMathingProfile = append(candidatesWithoutMathingProfile, &seed)
+	}
+
+	if len(candidatesWithMathingProfile) != 0 {
+		return candidatesWithMathingProfile[0], nil
+	}
+
+	if len(candidatesWithoutMathingProfile) != 0 {
+		return candidatesWithoutMathingProfile[0], nil
+	}
+	return nil, fmt.Errorf("failed to find valid seed for scheduling")
+}
+
+func verifySeedAvailability(seed *gardenv1beta1.Seed) bool {
+	if cond := gardencorev1alpha1helper.GetCondition(seed.Status.Conditions, gardenv1beta1.SeedAvailable); cond != nil {
+		return cond.Status == gardencorev1alpha1.ConditionTrue
+	}
+	return false
+}
+
+// updateBackupBucketToBeScheduledOntoSeed sets the seed name where the backupBucket should be scheduled on. Then it executes the actual update call to the API server. The call is capsuled to allow for easier testing.
+func (r *reconciler) updateBackupBucketToBeScheduledOntoSeed(backupBucket *gardencorev1alpha1.BackupBucket, seedName string) error {
+	return kutil.TryUpdate(r.ctx, retry.DefaultBackoff, r.client, backupBucket, func() error {
+		if backupBucket.Spec.Seed != nil {
+			alreadyScheduledErr := common.NewAlreadyScheduledError(fmt.Sprintf("backupBucket has already a seed assigned when trying to schedule the backupBucket to %s", *backupBucket.Spec.Seed))
+			return &alreadyScheduledErr
+		}
+		backupBucket.Spec.Seed = &seedName
+		return nil
+	})
+}
+
+func (r *reconciler) reportFailedScheduling(backupBucket *gardencorev1alpha1.BackupBucket, err error) {
```

Comment:
Not sure if such a function that only does one thing is really worth it.. :) your decision